### ISSUE TITLE
Breaking change in highlighting of code under a proc-macro

### ISSUE THREAD

**Issue by Veetaha** [rust-lang/rust-analyzer#18438]
This breaking change happened during the transition from version `v0.4.2159 -> v0.4.2160`. I tested this by switching between different "Pre-release" versions in VSCode:
![Image](https://github.com/user-attachments/assets/ceaece4f-9784-421f-84f8-1a1c7eff110b)

**code snippet to reproduce**: 

Add this proc macro as a dependency (using a `master` git dependency is required):
```toml
bon = { git = "https://github.com/elastio/bon" }
```

```rust
#[bon::builder]
fn example(arg: u32) {
    let _ = arg;
}
```

Code highlighting in `v0.4.2159` works correctly:

<img src="https://github.com/user-attachments/assets/ad5f9150-f61e-400d-96ef-2e7f6c758cb9" width="200px"/>

Code highlighting in `v0.4.2160` is broken:

<img src="https://github.com/user-attachments/assets/2ec6a6cd-d9ac-42a3-adf9-218ec6ff563d" width="200px"/>

The problem is that the `arg` symbol is assigned the semantic token type of an `enum`:

<img src="https://github.com/user-attachments/assets/4f110f5c-d986-412f-b81b-39d593313305" width="400px"/>

---

Internally, the macro generates the following code, where the `arg` symbol is used as a name for an enum, however, that `enum` declaration appears syntactically higher in code:

```rust
mod example_builder {
    mod members {
        pub(in super::super) enum arg {}
    }
}

fn __orig_example(arg: u32) {
    let _ = arg;
}
```

 Previously I relied on the fact that rust-analyzer assigns the semantic token type to the symbol based on the last occurence of its span in the generated code, so I intentionally put the `fn __orig_example(arg: u32) { /**/ } ` at the end of the proc macro out so I'm sure RA doesn't mess up the syntax highlighting, because the original function is always at the end thus spans of its symbols will guarantee correct highlighting.

**Comment by ChayimFriedman2** [comment]
This isn't exactly a bug, we don't guarantee anything about token classification, it's all heuristics and it's very possible they have changed lately. This can be a question, though.

**Comment by Veetaha** [comment]
Indeed, I guess the question is "how can I ensure consistent highlighting"?

**Comment by Veetaha** [comment]
I think I've solved this by reordering the generated items. It seems like now the rule is that "the first occurrence of the span wins" rather than "the last". I just want to make this doesn't change in the future and users of `bon` don't suddenly see broken highlighting

**Comment by ChayimFriedman2** [comment]
I don't think you can guarantee that, nor I think we want to guarantee that. If we will find a better heuristic, we will use it, and we don't want to be limited by previous guarantees. You can either accept that your macro may show the incorrect semantic highlighting, or try to follow changes here, but still accept that for some times, user may see incorrect highlighting.

**Comment by Veykril** [comment]
Yes we don't guarantee the order her, but we shouldn't have a reason to change it again in my eyes (in order to me sounds more like what people would expect instead of reverse order) either. Ideally we'd have

> A permanent fix for that may be to have some explicit way for macros to override token classification.

but that design space to be explored (and also commitment to exposed API which is always tough.

**Comment by ChayimFriedman2** [comment]
I wonder if the lang team will accept additional API to `proc_macro` that takes a span and marks it with a classification (from a `#[non_exhaustive]` enum).

If not we can probably inject it somehow it the resulting code, but it will be uglier.

**Comment by Veykril** [comment]
> Would emitting a custom (tool) attribute in front of the "main" usage in the token stream work?
> 
> mod example_builder {
>     mod members {
>         pub(in super::super) enum arg {}
>     }
> }
> 
> fn __orig_example(#[rust_analyzer::main_usage] arg: u32) {
>     let _ = arg;
> }
> 
> or something like that.

A tool attribute is what I would imagine as well. I don't think the proc-macro API wants to bother with this (the proc-macro implementation is already really complicated and the compiler wouldn't benefit from this whatsoever)

**Comment by davidbarsky** [comment]
I think it might be useful to spin up an issue tracker along the lines of the existing client-facing changes issue tracker, but aimed specifically at proc macro authors? Tangentially related, should rust-analyzer consider something like https://github.com/intellij-rust/intellij-rust/pull/9711?

**Comment by lnicola** [comment]
https://github.com/search?type=code&q=RUST_IDE_PROC_MACRO_COMPLETION, btw
